src/client/Client.cc

```diff
@@ -8556,9 +8564,13 @@ int Client::readdir_r_cb(dir_result_t *d, add_dirent_cb_t cb, void *p,
       in = diri->get_first_parent()->dir->parent_inode;
 
     int r;
+    if (in->is_dir()) {
+      caps |= CEPH_STAT_RSTAT;
+    }
     r = _getattr(in, caps, dirp->perms);
     if (r < 0)
       return r;
+    caps &= ~CEPH_STAT_RSTAT;
 
```

Comment:
Ditto here. '..' will always be a directory.

This looks like it'll fix '.' and '..', but what of the entries in for loop below this, around line 8625?

Reply:
".." direcotry  check have removed.
The entries in for loop below "." and ".." is aroud line 8445 , because in Client::_readdir_cache_cb，if the entries is readdir cache hit and it have all of the caps in the mask，it will never update it's rstat。

Reply:
If the inode turns out to not be in cache, and you do have to do the getattr around line 8445, don't you need to do the _getattr call with CEPH_CAP_RSTAT set in the mask?

Reply:
Have do _getattr call with CEPH_CAP_RSTAT set in the mask aroud line 8633, thanks!

Reply:
Have do _getattr call with CEPH_CAP_RSTAT set in the mask aroud line 8633, thanks!